File: ios/Classes/ImgEdit/ImageEditEngine.swift
import Flutter
import Foundation
import ImageIO
import MetalKit
import UIKit

/// Per-viewId editor session — owns the [MTKView], the [LayerStack], and
/// the [HistoryStack]. Method calls arrive routed by viewId from
/// `ImageEditPluginRouter`.
///
/// Real Metal pipeline lives in `Metal/*.metal` shaders + a `pipeline.swift`
/// that orchestrates `MTLRenderCommandEncoder` passes. Until the shaders
/// land, rendering is a no-op and the surface presents the source image
/// via a `CALayer` placeholder.
public final class ImageEditEngine: NSObject {
    public let viewId: Int64
    let mtkView: MTKView
    private let layers = LayerStack()
    private let history = HistoryStack()
    private var sourcePath: String?
    private var width: Int = 0
    private var height: Int = 0
    private let device: MTLDevice? = MTLCreateSystemDefaultDevice()
    private var renderer: ImageEditMetalRenderer?

    init(viewId: Int64, frame: CGRect, sourcePath: String?) {
        self.viewId = viewId
        self.mtkView = MTKView(frame: frame, device: device)
        self.sourcePath = sourcePath
        super.init()
        self.mtkView.framebufferOnly = false
        self.mtkView.colorPixelFormat = .bgra8Unorm
        self.mtkView.isPaused = true
        self.mtkView.enableSetNeedsDisplay = true
        if let device = device {
            let r = ImageEditMetalRenderer(device: device, viewId: viewId)
            r.hostView = self.mtkView
            self.renderer = r
            self.mtkView.delegate = r
        }
        if let path = sourcePath {
            renderer?.loadBaseImage(path: path)
        }
    }

    func detach() {
        renderer?.dispose()
        renderer = nil
        ImageEditPluginRouter.shared.unregister(viewId: viewId)
    }

    func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = (call.arguments as? [String: Any]) ?? [:]
        switch call.method {
        case "init":
            result(handleInit(args: args))
        case "dispose":
            detach()
            result(nil)
        case "pushLayer":
            let type = (args["type"] as? String) ?? "adjustment"
            let params = (args["params"] as? [String: Any]) ?? [:]
            let id = layers.push(type: type, params: params)
            history.recordPush(id: id, type: type, params: params)
            renderer?.setNeedsRender()
            result(["layerId": id])
        case "updateLayer":
            let id = (args["layerId"] as? Int) ?? -1
            let params = (args["params"] as? [String: Any]) ?? [:]
            let prev = layers.update(id: id, params: params)
            history.recordUpdate(id: id, prev: prev, next: params)
            renderer?.setNeedsRender()
            result(nil)
        case "removeLayer":
            let id = (args["layerId"] as? Int) ?? -1
            if let removed = layers.remove(id: id) {
                history.recordRemove(id: id, layer: removed)
            }
            renderer?.setNeedsRender()
            result(nil)
        case "reorderLayer":
            let id = (args["layerId"] as? Int) ?? -1
            let index = (args["index"] as? Int) ?? 0
            let from = layers.reorder(id: id, target: index)
            history.recordReorder(id: id, from: from, to: index)
            renderer?.setNeedsRender()
            result(nil)
        case "applyAdjust":
            let key = (args["key"] as? String) ?? ""
            let value = (args["value"] as? Double) ?? 0
            history.recordAdjust(key: key, prev: layers.currentAdjust(key), next: value)
            layers.applyAdjust(key: key, value: value)
            renderer?.applyAdjust(key: key, value: value)
            result(nil)
        case "applyFilter":
            let lutPath = (args["lutPath"] as? String) ?? ""
            let intensity = (args["intensity"] as? Double) ?? 1
            history.recordFilter(
                prevLut: layers.currentLut(),
                prevIntensity: layers.currentLutIntensity(),
                nextLut: lutPath,
                nextIntensity: intensity
            )
            layers.applyFilter(lutPath: lutPath, intensity: intensity)
            renderer?.applyFilter(lutPath: lutPath, intensity: intensity)
            result(nil)
        case "applyCrop":
            let rect = (args["rect"] as? [String: Any]) ?? [:]
            let rotation = (args["rotationDeg"] as? Double) ?? 0
            let persp = args["persp"] as? [Double]
            history.recordCrop(prev: layers.currentCrop(), rect: rect, rotationDeg: rotation, persp: persp)
            layers.applyCrop(rect: rect, rotationDeg: rotation, persp: persp)
            renderer?.applyCrop(rect: rect, rotationDeg: rotation, persp: persp)
            result(nil)
        case "brushStroke":
            renderer?.brushStroke(args: args)
            history.recordBlob(tag: "stroke", args: args)
            result(nil)
        case "spotHeal":
            renderer?.spotHeal(args: args)
            history.recordBlob(tag: "heal", args: args)
            result(nil)
        case "liquify":
            renderer?.liquify(args: args)
            history.recordBlob(tag: "liquify", args: args)
            result(nil)
        case "beautify":
            renderer?.beautify(args: args)
            history.recordBlob(tag: "beautify", args: args)
            result(nil)
        case "removeBg":
            var maskHandle = FaceDetector.runPersonSegmentation(
                viewId: viewId, sourcePath: sourcePath
            )
            if let dev = device, let path = sourcePath {
                let mask = BackgroundRemover.produceMaskTexture(
                    device: dev, sourcePath: path)
                if let tex = mask.texture {
                    renderer?.setMaskTexture(tex)
                    maskHandle["maskW"] = mask.width
                    maskHandle["maskH"] = mask.height
                    maskHandle["status"] = "ok"
                }
            }
            let id = layers.pushMask(maskHandle)
            history.recordPush(id: id, type: "mask", params: [:])
            result(["maskLayerId": id])
        case "placeText":
            let id = layers.push(type: "text", params: args)
            history.recordPush(id: id, type: "text", params: args)
            renderer?.setNeedsRender()
            result(["layerId": id])
        case "placeSticker":
            let id = layers.push(type: "sticker", params: args)
            history.recordPush(id: id, type: "sticker", params: args)
            renderer?.setNeedsRender()
            result(["layerId": id])
        case "placeEmoji":
            let id = layers.push(type: "emoji", params: args)
            history.recordPush(id: id, type: "emoji", params: args)
            renderer?.setNeedsRender()
            result(["layerId": id])
        case "undo":
            history.undo(into: layers)
            renderer?.setNeedsRender()
            result(["canUndo": history.canUndo, "canRedo": history.canRedo])
        case "redo":
            history.redo(into: layers)
            renderer?.setNeedsRender()
            result(["canUndo": history.canUndo, "canRedo": history.canRedo])
        case "setHistoryCap":
            let mb = (args["mb"] as? Int) ?? 64
            history.setMemoryCap(mb * 1024 * 1024)
            result(["capBytes": history.memoryCapBytes()])
        case "exportImage":
            let format = (args["format"] as? String) ?? "jpeg"
            let quality = (args["quality"] as? Int) ?? 92
            let maxDim = args["maxDim"] as? Int
            let path = (args["path"] as? String) ?? ""
            let res = Exporter.write(
                viewId: viewId,
                sourcePath: sourcePath,
                renderer: renderer,
                layers: layers,
                format: format,
                quality: quality,
                maxDim: maxDim,
                path: path
            )
            result(res)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleInit(args: [String: Any]) -> [String: Any] {
        let path = args["sourcePath"] as? String
        sourcePath = path
        let (w, h, exif) = decodeBounds(path: path)
        width = w
        height = h
        renderer?.initSession(sourcePath: path, width: w, height: h)
        layers.reset(width: w, height: h)
        history.clear()
        ImageEditPluginRouter.shared.notifyState(
            kind: "init",
            payload: ["viewId": viewId, "w": w, "h": h]
        )
        return [
            "viewId": viewId,
            "w": w,
            "h": h,
            "exif": exif,
        ]
    }

    private func decodeBounds(path: String?) -> (Int, Int, [String: Any]) {
        guard let path = path,
              let src = CGImageSourceCreateWithURL(
                URL(fileURLWithPath: path) as CFURL, nil)
        else {
            return (0, 0, [:])
        }
        guard let props = CGImageSourceCopyPropertiesAtIndex(src, 0, nil)
                as? [CFString: Any]
        else {
            return (0, 0, [:])
        }
        let w = (props[kCGImagePropertyPixelWidth] as? Int) ?? 0
        let h = (props[kCGImagePropertyPixelHeight] as? Int) ?? 0
        var exif: [String: Any] = [:]
        if let exifDict = props[kCGImagePropertyExifDictionary] as? [String: Any] {
            exif = exifDict
        }
        return (w, h, exif)
    }
}

private extension LayerStack {
    func currentAdjust(_ key: String) -> Double { adjustValue(forKey: key) }
}

File: android/src/main/kotlin/com/loopit/minis/imgedit/ImageEditEngine.kt
package com.loopit.minis.imgedit

import android.graphics.BitmapFactory
import android.util.Log
import android.view.SurfaceHolder
import android.view.SurfaceView
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.util.concurrent.atomic.AtomicInteger

/**
 * Per-viewId session — owns the GLES surface lifecycle, the [LayerStack],
 * and the [HistoryStack]. Channel calls arrive routed by viewId from
 * [ImageEditPluginRouter].
 *
 * Render graph (see `cpp/imgedit/pipeline.cpp`):
 *   BaseImageLayer → Adjustments → LUT → per-layer composite → Output FBO.
 */
class ImageEditEngine private constructor(
    val viewId: Int,
    private val surface: SurfaceView,
) : SurfaceHolder.Callback {

    private val tag = "MinisImgEdit"
    private val layers = LayerStack()
    private val history = HistoryStack()
    private var sourcePath: String? = null
    private var width: Int = 0
    private var height: Int = 0

    init {
        surface.holder.addCallback(this)
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        ImageEditNative.surfaceCreated(viewId, holder.surface)
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, w: Int, h: Int) {
        ImageEditNative.surfaceChanged(viewId, w, h)
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        ImageEditNative.surfaceDestroyed(viewId)
    }

    fun detach() {
        try {
            ImageEditNative.disposeView(viewId)
            ImageEditPluginRouter.unregister(viewId)
        } catch (t: Throwable) {
            Log.w(tag, "detach failed: ${t.message}")
        }
    }

    fun handle(call: MethodCall, result: MethodChannel.Result) {
        val args = (call.arguments as? Map<*, *>) ?: emptyMap<Any, Any?>()
        when (call.method) {
            "init" -> result.success(handleInit(args))
            "dispose" -> {
                detach()
                result.success(null)
            }
            "pushLayer" -> {
                val type = args["type"] as? String ?: "adjustment"
                @Suppress("UNCHECKED_CAST")
                val params = (args["params"] as? Map<String, Any?>) ?: emptyMap()
                val id = layers.push(type, params)
                history.recordPush(id, type, params)
                ImageEditNative.requestRender(viewId)
                result.success(mapOf("layerId" to id))
            }
            "updateLayer" -> {
                val id = (args["layerId"] as? Number)?.toInt() ?: -1
                @Suppress("UNCHECKED_CAST")
                val params = (args["params"] as? Map<String, Any?>) ?: emptyMap()
                val prev = layers.update(id, params)
                history.recordUpdate(id, prev, params)
                ImageEditNative.requestRender(viewId)
                result.success(null)
            }
            "removeLayer" -> {
                val id = (args["layerId"] as? Number)?.toInt() ?: -1
                val removed = layers.remove(id)
                if (removed != null) history.recordRemove(id, removed)
                ImageEditNative.requestRender(viewId)
                result.success(null)
            }
            "reorderLayer" -> {
                val id = (args["layerId"] as? Number)?.toInt() ?: -1
                val index = (args["index"] as? Number)?.toInt() ?: 0
                val from = layers.reorder(id, index)
                history.recordReorder(id, from, index)
                ImageEditNative.requestRender(viewId)
                result.success(null)
            }
            "applyAdjust" -> {
                val key = args["key"] as? String ?: ""
                val value = (args["value"] as? Number)?.toDouble() ?: 0.0
                history.recordAdjust(key, layers.currentAdjust(key), value)
                layers.applyAdjust(key, value)
                ImageEditNative.applyAdjust(viewId, key, value)
                result.success(null)
            }
            "applyFilter" -> {
                val lutPath = args["lutPath"] as? String ?: ""
                val intensity = (args["intensity"] as? Number)?.toDouble() ?: 1.0
                history.recordFilter(layers.currentLut(), layers.currentLutIntensity(), lutPath, intensity)
                layers.applyFilter(lutPath, intensity)
                ImageEditNative.applyFilter(viewId, lutPath, intensity)
                result.success(null)
            }
            "applyCrop" -> {
                @Suppress("UNCHECKED_CAST")
                val rect = (args["rect"] as? Map<String, Any?>) ?: emptyMap()
                val rotation = (args["rotationDeg"] as? Number)?.toDouble() ?: 0.0
                @Suppress("UNCHECKED_CAST")
                val persp = args["persp"] as? List<Double>
                history.recordCrop(layers.currentCrop(), rect, rotation, persp)
                layers.applyCrop(rect, rotation, persp)
                ImageEditNative.applyCrop(viewId, rect, rotation, persp)
                result.success(null)
            }
            "brushStroke" -> {
                ImageEditNative.brushStroke(viewId, args)
                history.recordStroke(args)
                result.success(null)
            }
            "spotHeal" -> {
                ImageEditNative.spotHeal(viewId, args)
                history.recordHeal(args)
                result.success(null)
            }
            "liquify" -> {
                ImageEditNative.liquify(viewId, args)
                history.recordLiquify(args)
                result.success(null)
            }
            "beautify" -> {
                ImageEditNative.beautify(viewId, args)
                history.recordBeautify(args)
                result.success(null)
            }
            "removeBg" -> {
                val maskLayer = FaceDetector.runSelfieSegmentation(viewId, sourcePath)
                val id = layers.pushMask(maskLayer)
                history.recordPush(id, "mask", emptyMap())
                result.success(mapOf("maskLayerId" to id))
            }
            "placeText" -> {
                val params = args.castMap()
                val id = layers.push("text", params)
                history.recordPush(id, "text", params)
                ImageEditNative.requestRender(viewId)
                result.success(mapOf("layerId" to id))
            }
            "placeSticker" -> {
                val params = args.castMap()
                val id = layers.push("sticker", params)
                history.recordPush(id, "sticker", params)
                ImageEditNative.requestRender(viewId)
                result.success(mapOf("layerId" to id))
            }
            "placeEmoji" -> {
                val params = args.castMap()
                val id = layers.push("emoji", params)
                history.recordPush(id, "emoji", params)
                ImageEditNative.requestRender(viewId)
                result.success(mapOf("layerId" to id))
            }
            "undo" -> {
                history.undo(layers)
                ImageEditNative.requestRender(viewId)
                result.success(mapOf("canUndo" to history.canUndo(), "canRedo" to history.canRedo()))
            }
            "redo" -> {
                history.redo(layers)
                ImageEditNative.requestRender(viewId)
                result.success(mapOf("canUndo" to history.canUndo(), "canRedo" to history.canRedo()))
            }
            "setHistoryCap" -> {
                val mb = (args["mb"] as? Number)?.toLong() ?: 64L
                history.setMemoryCap(mb * 1024L * 1024L)
                result.success(mapOf("capBytes" to history.memoryCap()))
            }
            "exportImage" -> {
                val format = args["format"] as? String ?: "jpeg"
                val quality = (args["quality"] as? Number)?.toInt() ?: 92
                val maxDim = (args["maxDim"] as? Number)?.toInt()
                val path = args["path"] as? String ?: ""
                val exported = Exporter.write(
                    viewId = viewId,
                    sourcePath = sourcePath,
                    layers = layers,
                    format = format,
                    quality = quality,
                    maxDim = maxDim,
                    path = path,
                )
                result.success(exported)
            }
            else -> result.notImplemented()
        }
    }

    private fun handleInit(args: Map<*, *>): Map<String, Any?> {
        val path = args["sourcePath"] as? String
        sourcePath = path
        val (w, h, exif) = decodeBounds(path)
        width = w
        height = h
        ImageEditNative.initSession(viewId, path, w, h)
        layers.reset(w, h)
        history.clear()
        ImageEditPluginRouter.notifyState(
            "init",
            mapOf("viewId" to viewId, "w" to w, "h" to h),
        )
        return mapOf(
            "viewId" to viewId,
            "w" to w,
            "h" to h,
            "exif" to exif,
        )
    }

    private fun decodeBounds(path: String?): Triple<Int, Int, Map<String, Any?>> {
        if (path.isNullOrEmpty()) return Triple(0, 0, emptyMap())
        val f = File(path)
        if (!f.exists()) return Triple(0, 0, emptyMap())
        val opts = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(path, opts)
        return Triple(
            opts.outWidth.coerceAtLeast(0),
            opts.outHeight.coerceAtLeast(0),
            emptyMap(),
        )
    }

    companion object {
        private val NEXT_LAYER_ID = AtomicInteger(1)

        fun create(viewId: Int, surface: SurfaceView, sourcePath: String?): ImageEditEngine {
            val engine = ImageEditEngine(viewId, surface)
            engine.sourcePath = sourcePath
            return engine
        }

        fun nextLayerId(): Int = NEXT_LAYER_ID.getAndIncrement()
    }
}

private fun Map<*, *>.castMap(): Map<String, Any?> {
    val out = HashMap<String, Any?>(size)
    for ((k, v) in this) {
        if (k is String) out[k] = v
    }
    return out
}
